src/mito2/src/cache/write_cache.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A write-through cache for remote object stores.
+
+use std::sync::Arc;
+
+use object_store::manager::ObjectStoreManagerRef;
+use object_store::ObjectStore;
+use store_api::metadata::RegionMetadataRef;
+use store_api::storage::{RegionId, SequenceNumber};
+use tokio::sync::mpsc::Sender;
+
+use crate::access_layer::sst_file_path;
+use crate::error::Result;
+use crate::read::Source;
+use crate::request::WorkerRequest;
+use crate::sst::file::{FileId, FileMeta, Level};
+use crate::sst::parquet::writer::ParquetWriter;
+use crate::sst::parquet::WriteOptions;
+use crate::wal::EntryId;
+
+/// A cache for uploading files to remote object stores.
+///
+/// It keeps files in local disk and then sends files to object stores.
+pub(crate) struct WriteCache {
+    /// Local object storage to store files to upload.
+    local_store: ObjectStore,
+    /// Object store manager.
+    object_store_manager: ObjectStoreManagerRef,
+}
+
+pub(crate) type WriteCacheRef = Arc<WriteCache>;
+
+impl WriteCache {
+    // TODO(yingwen): Maybe pass cache path instead of local store.
+    /// Create the cache with a `local_store` to cache files and a
+    /// `object_store_manager` for all object stores.
+    pub(crate) fn new(
+        local_store: ObjectStore,
+        object_store_manager: ObjectStoreManagerRef,
+    ) -> Self {
+        // TODO(yingwen): Cache capacity.
+        Self {
+            local_store,
+            object_store_manager,
+        }
+    }
+
+    /// Adds files to the cache.
+    pub(crate) async fn upload(&self, upload: Upload) -> Result<()> {
+        // Add the upload metadata to the manifest.
+        unimplemented!()
+    }
+}
+
+/// A remote write request to upload files.
+pub(crate) struct Upload {
+    /// Parts to upload.
+    pub(crate) parts: Vec<UploadPart>,
+}
+
+/// Metadata of SSTs to upload together.
+pub(crate) struct UploadPart {
```

Comment:
If the `UploadPart` only serves for SSTs, can I still have the chance to write files of other types in the parquet writer?

Reply:
Good question. Now it only serves for SSTs, we might store the file type in the `FileMeta` to support other files.